File: src/main/java/com/goudourasv/http/instructors/InstructorsResource.java
package com.goudourasv.http.instructors;

import com.goudourasv.http.instructors.dto.InstructorCreate;
import com.goudourasv.http.instructors.dto.InstructorUpdate;
import com.goudourasv.domain.instructors.Instructor;
import com.goudourasv.domain.instructors.InstructorsService;

import javax.validation.Valid;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.UriInfo;
import java.net.URI;
import java.util.List;
import java.util.UUID;

@Path("/instructors")
public class InstructorsResource {
    private final InstructorsService instructorsService = new InstructorsService();

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    public List<Instructor> getInstructors(@QueryParam("institution") String institution) {
        List<Instructor> instructorsList = instructorsService.getInstructors();
        List<Instructor> filteredInstructorsList = instructorsService.getFilteredInstructors(instructorsList, institution);
        return instructorsList;
    }

    @GET
    @Path("{id}")
    @Produces(MediaType.APPLICATION_JSON)
    public Instructor getSpecificInstructor(@PathParam("id") UUID id) {
        Instructor specificInstructor = instructorsService.getSpecificInstructor(id);
        return specificInstructor;
    }

    @POST
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    public Response createInstructor(@Valid InstructorCreate input, UriInfo uriInfo) {
        Instructor createdInstructor = instructorsService.createNewInstructor(input);

        String path = uriInfo.getPath();
        String location = path + createdInstructor.getId().toString();
        return Response.created(URI.create(location)).entity(createdInstructor).build();
    }

    @DELETE
    @Path("{id}")
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    public void deleteSpecificInstructor(@PathParam("id") UUID id) {
        instructorsService.deleteSpecificInstructor(id);
    }

    @PUT
    @Path("{id}")
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    public Instructor updateInstructor(@PathParam("id") UUID id, @Valid InstructorCreate instructor) {
        Instructor updatedInstructor = instructorsService.replaceInstructor(id, instructor);
        return updatedInstructor;
    }

    @PATCH
    @Path("/{id}")
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    public Instructor partiallyUpdateInstructor(@PathParam("id") UUID id, InstructorUpdate input) {
        Instructor instructorToUpdate = instructorsService.partiallyUpdateInstructor(input, id);
        return instructorToUpdate;
    }


}

File: src/main/java/com/goudourasv/http/instructors/InstructorsResource.kt
package com.goudourasv.http.instructors

import com.goudourasv.domain.instructors.Instructor
import com.goudourasv.domain.instructors.InstructorsService
import com.goudourasv.http.instructors.dto.InstructorCreate
import com.goudourasv.http.instructors.dto.InstructorUpdate
import io.smallrye.common.annotation.Blocking
import java.net.URI
import java.util.*
import javax.enterprise.context.ApplicationScoped
import javax.validation.Valid
import javax.ws.rs.*
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response
import javax.ws.rs.core.UriInfo

@ApplicationScoped
@Path("/instructors")
class InstructorsResource(private val instructorsService: InstructorsService) {

    @Blocking
    @GET
    @Produces(MediaType.APPLICATION_JSON)
    fun getInstructors(@QueryParam("institution") institutionId: UUID?): List<Instructor> {
        return instructorsService.getInstructors(institutionId)

    }

    @Blocking
    @GET
    @Path("/{id}")
    @Produces(MediaType.APPLICATION_JSON)
    fun getSpecificInstructor(@PathParam("id") instructorId: UUID): Instructor {
        return instructorsService.getSpecificInstructor(instructorId) ?: throw NotFoundException()
    }

    @Blocking
    @POST
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    fun createInstructor(@Valid instructorCreate: InstructorCreate, uriInfo: UriInfo): Response {
        val createdInstructor = instructorsService.createNewInstructor(instructorCreate)
        val path: String = uriInfo.path
        val location: String = path + createdInstructor.id
        return Response.created(URI.create(location)).entity(createdInstructor).build()
    }

    @Blocking
    @DELETE
    @Path("/{id}")
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    fun deleteInstructor(@PathParam("id") instructorId: UUID) {
        val deleted = instructorsService.deleteSpecificInstructor(instructorId)
        if (!deleted) {
            throw  NotFoundException("Instructor with id $instructorId doesn't exist")
        }
    }

    @Blocking
    @PUT
    @Path("/{id}")
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    fun updateInstructor(@PathParam("id") instructorId: UUID, @Valid instructorCreate: InstructorCreate): Instructor {
        return instructorsService.replaceInstructor(instructorId, instructorCreate)

    }

    @Blocking
    @PATCH
    @Path("/{id}")
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    fun partiallyUpdateInstructor(@PathParam("id") instructorId: UUID, instructorUpdate: InstructorUpdate): Instructor {
        return instructorsService.partiallyUpdateInstructor(instructorId, instructorUpdate)
    }

}